### ISSUE TITLE
[Paper Note] Better & Faster Large Language Models via Multi-token Prediction, Fabian Gloeckle+, ICML'24

### ISSUE THREAD

**Issue by AkihikoWatanabe** [AkihikoWatanabe/paper_notes#2444]
# URL
- https://arxiv.org/abs/2404.19737
# Authors
  - Fabian Gloeckle
  - Badr Youbi Idrissi
  - Baptiste Rozière
  - David Lopez-Paz
  - Gabriel Synnaeve
# Abstract
  - Large language models such as GPT and Llama are trained with a next-token prediction loss. In this work, we suggest that training language models to predict multiple future tokens at once results in higher sample efficiency. More specifically, at each position in the training corpus, we ask the model to predict the following n tokens using n independent output heads, operating on top of a shared model trunk. Considering multi-token prediction as an auxiliary training task, we measure improved downstream capabilities with no overhead in training time for both code and natural language models. The method is increasingly useful for larger model sizes, and keeps its appeal when training for multiple epochs. Gains are especially pronounced on generative benchmarks like coding, where our models consistently outperform strong baselines by several percentage points. Our 13B parameter models solves 12 % more problems on HumanEval and 17 % more on MBPP than comparable next-token models. Experiments on small algorithmic tasks demonstrate that multi-token prediction is favorable for the development of induction heads and algorithmic reasoning capabilities. As an additional benefit, models trained with 4-token prediction are up to 3 times faster at inference, even with large batch sizes.
# Translation (by gpt-4o-mini)
- 大規模言語モデル（GPTやLlamaなど）は、次のトークン予測損失を用いて訓練されています。本研究では、言語モデルを同時に複数の将来のトークンを予測するように訓練することで、サンプル効率が向上することを提案します。具体的には、訓練コーパスの各位置において、モデルにn個の独立した出力ヘッドを使用して次のnトークンを予測させ、共有モデルのトランクの上で動作させます。多トークン予測を補助的な訓練タスクと見なすことで、コードおよび自然言語モデルの両方において、訓練時間にオーバーヘッドをかけることなく、下流の能力が向上することを測定しました。この手法は、モデルサイズが大きくなるにつれてますます有用であり、複数のエポックで訓練する際にもその魅力を保ちます。特に、コーディングのような生成ベンチマークでは、我々のモデルが強力なベースラインを数パーセントポイント上回ることが顕著です。我々の13Bパラメータモデルは、HumanEvalで12%多くの問題を解決し、MBPPでは17%多くの問題を解決します。小規模なアルゴリズムタスクに関する実験は、多トークン予測が誘導ヘッドやアルゴリズム的推論能力の発展に有利であることを示しています。さらに、4トークン予測で訓練されたモデルは、大きなバッチサイズでも推論時に最大3倍速くなります。
# Summary (by gpt-4o-mini)
- 本研究では、大規模言語モデルを複数の将来のトークンを同時に予測するように訓練する手法を提案し、サンプル効率の向上を図る。具体的には、n個の独立した出力ヘッドを用いて次のnトークンを予測し、訓練時間にオーバーヘッドをかけずに下流の能力を向上させる。特に、コーディングタスクにおいて、提案モデルは強力なベースラインを上回る性能を示し、推論時に最大3倍の速度向上も実現。

**Comment by AkihikoWatanabe** [comment]
next tokenだけでなく、next 4-tokenを予測して学習することで、MBPP/HumanEvalにおいて、モデルのパラメータサイズが1.3Bを超えた時点でベースライン（=同じパラメータサイズとなるように調整されたnext-token prediction）をoutperformしはじめ、モデルサイズが大きくなるにつれて性能の差が顕著に表れることを示した。コーディングドメインにおいて事前学習、およびfinetuningの双方で効果がある。ただし、3.7節で示されている通り、これはコーディングドメインでのみこのような顕著な改善がみられており、自然言語データに対してはここまで顕著な改善はしていないように見える（5.1節で考察されていそう; 昨今のLLMでは事前学習データにコーディングなどのデータが入るのが普通なので利用する恩恵はありそう; Abstractive Summarizationでは性能が改善している(Figure6); GSM8Kでは200Bまではnext 2 tokenを予測すると性能が改善しているが500B token学習するとnext token predictionの方が性能が良くなる）。全体的にperplexityの改善（=次のトークンにおいて正解トークンの生成確率を改善する）というよりは、モデルの"最終的な生成結果”にフォーカスした評価となっている。

モデルは共有のトランクf_s (おそらくhead間でパラメータを共有している一連のtransformerブロック) を持っておりinput x_t:1に対応するlatent representation z_t:1を生成する。latent representationをoutput headにinputすることで、それぞれのheadが合計でn個のnext tokenを予測する。
<img width="608" height="1021" alt="Image" src="https://github.com/user-attachments/assets/433d69cb-5593-483b-b591-6445c482ed2e" />

next n-tokenを予測する際には、GPUメモリを大幅に食ってしまう （logitsのshapeが(n, V)となりそれらの勾配も保持しなければならない) ことがボトルネックとなるが、f_sまでforward passを実行したら、各headに対してforward/backward passを順番に実行して、logitsの値は破棄し勾配の情報だけf_sに蓄積することで、長期的に保持する情報を各headのから逆伝搬された勾配情報のみにすることでこれを解決している。
<img width="597" height="478" alt="Image" src="https://github.com/user-attachments/assets/3f5ff3fc-5934-4f12-9327-23b689526464" />

実際にinferenceをするときはnext tokenを予測するヘッドの出力を活用することを前提としているが、全てのヘッドを活用することで、t時点でt+nトークンの予測を可能なため、self-speculative decodingを実施しinference timeを短縮することができる。

3.4で示されているように、nの値は大きければ大きいほど良いというわけではなく、4程度（byte levelなモデルの場合は8 bytes）が最適なようである。が、Table1を見ると、データによってはn=6が良かったり（i.e., 最適なnは学習データ依存）複数エポック学習するとmulti token predictionの効果が薄くなっていそう（i.e., 同じトークンの予測を複数回学習するので実質multi token predictionと似たようなことをやっている。言い換えると、multi token predictionは複数epochの学習を先取りしているとみなせる？）なのは注意が必要そう。

**Comment by AkihikoWatanabe** [comment]
全体的に複数epochを学習すると恩恵がなくなっていく（コーディング） or next token predictionよりも性能が悪化する（自然言語）ので、LLMの事前学習において、複数epochを学習するような当たり前みたいな世界線が訪れたら、このアーキテクチャを採用すると性能はむしろ悪化しそうな気はする。

**Comment by AkihikoWatanabe** [comment]
MBPP/HumanEval:
- #2439 
- #2438
